lua/null-ls/builtins/formatting/nix_flake_fmt.lua

```diff
@@ -0,0 +1,145 @@
+        return nil
+    end
+    local nix_current_system = cp.stdout
+
+    local eval_nix_formatter = [[
+      let
+        currentSystem = ]] .. nix_current_system .. [[;
+        # Various functions vendored from nixpkgs lib (to avoid adding a
+        # dependency on nixpkgs).
+        lib = rec {
+          getOutput = output: pkg:
+            if ! pkg ? outputSpecified || ! pkg.outputSpecified
+            then pkg.${output} or pkg.out or pkg
+            else pkg;
+          getBin = getOutput "bin";
+          # Simplified by removing various type assertions.
+          getExe' = x: y: "${getBin x}/bin/${y}";
+          # getExe is simplified to assume meta.mainProgram is specified.
+          getExe = x: getExe' x x.meta.mainProgram;
+        };
+      in
+      formatterBySystem:
+        if formatterBySystem ? ${currentSystem} then
+          let
+            formatter = formatterBySystem.${currentSystem};
+            drv = formatter.drvPath;
+            bin = lib.getExe formatter;
+          in
+            drv + "\n" + bin + "\n"
+        else
+          ""
+    ]]
+
+    cp = vim.system({ "nix", "eval", ".#formatter", "--raw", "--apply", eval_nix_formatter }, { cwd = params.root })
+        :wait()
+    if cp.code ~= 0 then
+        -- Dirty hack to check if the flake actually defines a formatter.
+        --
+        -- I cannot for the *life* of me figure out a less hacky way of
+        -- checking if a flake defines a formatter. Things I've tried:
+        --
+        --   - `nix eval . --apply '...'`: This doesn't not give me the flake
+        --                                 itself, it gives me the default package.
+        --   - `builtins.getFlake`: Every incantation I've tried requires
+        --                          `--impure`, which has the performance downside described above.
+        --   - `nix flake show --json .`: This works, but it can be quite slow:
+        --                                we end up evaluating all outputs, which can take a while for
+        --                                `nixosConfigurations`.
+        if cp.stderr:find("error: flake .+ does not provide attribute .+ or 'formatter'") then
+            log:warn("this flake does not define a `nix fmt` entrypoint")
+        else
+            log:error(string.format("unable discover 'nix fmt' command. stderr: %s", cp.stderr))
+        end
+
+        return nil
+    end
+
+    if cp.stdout == "" then
+        log:warn("this flake does not define a formatter for your system: %s", nix_current_system)
+        return nil
+    end
+
+    -- stdout has 2 lines of output:
+    --  1. drv path
+    --  2. exe path
+    local drv_path, nix_fmt_path = cp.stdout:match("([^\n]+)\n([^\n]+)\n")
+
+    -- Build the derivation. This ensures that `nix_fmt_path` exists.
+    cp = vim.system({ "nix", "build", "--no-link", drv_path .. "^out" }):wait()
```

Comment:
Ideally you would have some out link here in case the result gets garbage collected by other things. Don't know enough about vim plugins to give advise how to keep a temporary directory.

Reply:
Yeah, I was sort of hoping to gloss over garbage collection. How does this work for `nix fmt` itself? It's not registering gc roots, right? Do that work because it always re-evaluates and rebuilds?

Perhaps I could implement this the same way? Basically, change the caching logic here to check if the file has disappeared, and if it has disappeared, just do a rebuild.

That would save me the hassle of finding a temp directory to put a gcroot in and stressing over cleaning it up.

Reply:
I actually don't know if it has effective protection between build and exec. Once it's executed there will be a gcroot based on the process.
However the time window will be much shorter than what we have here.
You cannot do any caching in long running processes reliable without gcroots. You would need to build it every time in this case.
People might have their vim open for hours/days/weeks and we don't want to break this use case.

Reply:
You would need to run `nix-store --verify-path` to check if the path is still valid.
And it would still be racy because of the time window between the verify and the exec.

Reply:
I've added the store root as you suggested using `vim.fn.tempname`. Good find!

The following questions are not relevant to the PR, I just want to learn more about nix.

> You would need to run `nix-store --verify-path` to check if the path is still valid, whenever you look up the cache.

This is breaking my mental model a bit. I thought it would be enough to just check if the executable exists, and if it exists then everything it depends on must exist (by the rules of garabage collection).

I've never used `nix-store --verify-path` before. Reading the man page, it *sounds* like it's a tool for detecting corruption in /nix/store, which I don't think we're worried about here.

> And it would still be racy because of the time window between the verify and the exec.

This is a small window though, right? Effectively the same as the race you hypothesized might exist in cppnix between build and exec.